### PR TITLE
Refactor Phabricator backend and file_info code

### PR DESCRIPTION
**Pull request by lguychard**
Fixes https://github.com/sourcegraph/sourcegraph/issues/3823, https://github.com/sourcegraph/sourcegraph/issues/5051

This took me a good minute to get to a PRable state, and I'm sorry it's such a mammoth PR 😞

Notable changes:
- **Phabricator `FileInfo` resolvers no longer return rev names**. This was the source of many bugs, as returning the name of a rev that only exists on a staging repo, and was auto-generated by arc, doesn't make sense when the staging repo is not synced to the Sourcegraph instance. The main drawbacks are that the `OpenDiffOnSourcegraph` links will always contain commit IDs and not rev names, and that the workspace roots exposed to extensions will always have undefined inputRevisions. I think this is an acceptable tradeoff considering it a) gets rid of bugs and b) gets rid of a lot of hacky code like this (this is just one example): https://sourcegraph.com/github.com/sourcegraph/sourcegraph@a3add8a6499095ad4b62536e6b37a7ccf661c449/-/blob/browser/src/libs/phabricator/util.tsx#L354-359 
- Functions that queried the Phabricator Conduit API have been DRYed, now use `rxjs.ajax()` instead of `fetch()`, and `QueryConduitHelper` is now injected so that it can be mocked in tests. This allowed me to add tests for all fileInfo resolvers by running them on page snapshots, and providing mock Sourcegraph and Phabricator API responses.
- Terminology:
    - `differentialID` -> `revisionID`: in Differential, a set of changes up for review is a *revision*, which is composed of one or more *diffs* (patches).
    - `leftDiffID` becomes `baseDiffID`, to clarify the use case of comparing two diffs in Phabricator. This is the best I could come up with, I explained it in comments in the code.
- `getPhabricatorState()` now returns a properly discriminated union. This allowed me to find errors where we were building incorrect states.
- `FileInfo` resolvers now throw an error if they receive an unexpected state (instead of silently never emitting).
- Many scraping functions now throw precise errors as early as possible instead of doing a `console.error()` or nothing at all and then returning `null`.

Test plan:
- New unit tests for all `FileInfo` resolvers.
- I have done some manual testing of all permutations I added as unit tests.
- I'm considering adding some e2e tests as well, but I will do this in a follow-up PR.

### CODE REVIEW

File: browser/src/libs/phabricator/__snapshots__/file_info.test.ts.snap [hunk 1]
```diff
@@ -0,0 +1,73 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`Phabricator file info resolveDiffFileInfo() Differential revision - comparing diffs - staging repo not synced 1`] = `
```
[the remark this answers is not in the document]
Comment (lguychard) [reply browser/src/libs/phabricator/__snapshots__/file_info.test.ts.snap]: Done.

File: browser/src/libs/phabricator/backend.tsx [hunk 2]
```diff
@@ -126,123 +109,70 @@ export async function getPhabricatorCSS(sourcegraphURL: string): Promise<string>
...
+export type QueryConduitHelper<T> = (endpoint: string, params: {}) => Observable<T>
 
-interface ConduitCommit {
-    fields: {
-        identifier: string
+export function queryConduit<T>(endpoint: string, params: {}): Observable<T> {
+    const form = createConduitRequestForm()
+    for (const [key, value] of Object.entries(params)) {
+        form.set(`params[${key}]`, JSON.stringify(value))
     }
```
[the remark this answers is not in the document]
Comment (lguychard) [reply browser/src/libs/phabricator/backend.tsx]: `queryConduit()` is the sole caller of `createConduitRequestForm()` (there was only one other useless callsite left in  `getRepoDetailsFromRepoPHID()`, which I just removed).

But yes, the `params[key], JSON.stringify()` was done at every callsite before 😞

File: browser/src/libs/phabricator/backend.tsx [hunk 3]
```diff
@@ -126,123 +109,70 @@ export async function getPhabricatorCSS(sourcegraphURL: string): Promise<string>
...
-interface ConduitDiffusionCommitQueryResponse {
-    error_code?: string
-    error_info?: string
-    result: {
-        data: ConduitCommit[]
-    }
+function getDiffDetailsFromConduit(
+    diffID: number,
+    revisionID: number,
+    queryConduit: QueryConduitHelper<ConduitDiffDetailsResponse>
```
[the remark this answers is not in the document]
Comment (lguychard) [reply browser/src/libs/phabricator/backend.tsx]: Great point. It was never called with literal values as argument, so I didn't spot it 🙂

File: browser/src/libs/phabricator/backend.tsx [hunk 4]
```diff
@@ -274,45 +204,35 @@ interface PhabricatorRepoDetails {
...
+        switchMap((details: PhabricatorRepoDetails | null) => {
+            if (!details) {
+                return throwError(new Error('could not parse repo details'))
+            }
+            return createPhabricatorRepo({
+                callsign,
+                repoName: details.rawRepoName,
+                phabricatorURL: window.location.origin,
+                requestGraphQL,
+            }).pipe(mapTo(details)) as Observable<PhabricatorRepoDetails>
```
[the remark this answers is not in the document]
Comment (lguychard) [reply browser/src/libs/phabricator/backend.tsx]: Not sure why I added them during development, they are indeed unneeded.

File: browser/src/libs/phabricator/backend.tsx [hunk 5]
```diff
@@ -126,123 +109,70 @@ export async function getPhabricatorCSS(sourcegraphURL: string): Promise<string>
...
+    )
 }
 
-interface ConduitDiffusionCommitQueryResponse {
-    error_code?: string
-    error_info?: string
-    result: {
-        data: ConduitCommit[]
-    }
+function getDiffDetailsFromConduit(
```
[the remark this answers is not in the document]
Comment (lguychard) [reply browser/src/libs/phabricator/backend.tsx]: Improved the docstrings, the most relevant/helpful one is that of `resolveDiffRev()`

File: browser/src/libs/phabricator/backend.tsx [hunk 6]
```diff
@@ -126,123 +109,70 @@ export async function getPhabricatorCSS(sourcegraphURL: string): Promise<string>
...
+        url: window.location.origin + endpoint,
+        method: 'POST',
+        body: form,
+        withCredentials: true,
+        headers: {
+            Accept: 'application/json',
+        },
+    }).pipe(
+        map(({ response }: { response: ConduitResponse<T> }) => {
+            if (response.error_code !== null) {
```
[the remark this answers is not in the document]
Comment (lguychard) [reply browser/src/libs/phabricator/backend.tsx]: Yes, Conduit returns 200 OK on errors, with a custom `error_code` (🙄). Example of a 200 OK response body:

```
{"result":null,"error_code":"ERR-CONDUIT-CORE","error_info":"Constraint \"callfoos\" is not a valid constraint for this query."}
```

File: browser/src/libs/phabricator/file_info.test.ts [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+
+interface Fixture {
+    htmlFixture: string
+    url: string
+    codeViewSelector: string
+    graphQLResponseMap?: GraphQLResponseMap
+    conduitResponseMap?: ConduitResponseMap
+}
+
+const getFileInfo = async (
```
Comment (felixfbecker) [inline browser/src/libs/phabricator/file_info.test.ts]: This reads like a pure function but modifies the DOM and URL. Is there a better name?

Comment (lguychard) [reply browser/src/libs/phabricator/file_info.test.ts]: Good call - renamed to `resolveFileInfoFromFixture()`

**Pull request merged**